src/renderer/atlas/AtlasEngine.r.cpp

```diff
@@ -0,0 +1,460 @@
+        wil::com_ptr<ID2D1SolidColorBrush> brush;
+        THROW_IF_FAILED(_r.d2dRenderTarget->CreateSolidColorBrush(&color, nullptr, brush.addressof()));
+        _r.brush = brush.query<ID2D1Brush>();
+    }
+
+    _r.scratchpadCellWidth = _r.maxEncounteredCellCount;
+    WI_SetAllFlags(_r.invalidations, RenderInvalidations::ConstBuffer);
+}
+
+void AtlasEngine::_processGlyphQueue()
+{
+    if (_r.glyphQueue.empty())
+    {
+        return;
+    }
+
+    for (const auto& pair : _r.glyphQueue)
+    {
+        _drawGlyph(pair);
+    }
+
+    _r.glyphQueue.clear();
+}
+
+void AtlasEngine::_drawGlyph(const AtlasQueueItem& item) const
+{
+    const auto key = item.key->data();
+    const auto value = item.value->data();
+    const auto coords = &value->coords[0];
+    const auto charsLength = key->charCount;
+    const auto cells = static_cast<u32>(key->attributes.cellCount);
+    const auto textFormat = _getTextFormat(key->attributes.bold, key->attributes.italic);
+
+    // See D2DFactory::DrawText
+    wil::com_ptr<IDWriteTextLayout> textLayout;
+    THROW_IF_FAILED(_sr.dwriteFactory->CreateTextLayout(&key->chars[0], charsLength, textFormat, cells * _r.cellSizeDIP.x, _r.cellSizeDIP.y, textLayout.addressof()));
+    if (item.scale != 1.0f)
+    {
+        const auto f = textFormat->GetFontSize();
+        textLayout->SetFontSize(f * item.scale, { 0, charsLength });
+    }
+    if (_r.typography)
+    {
+        textLayout->SetTypography(_r.typography.get(), { 0, charsLength });
+    }
+
+    auto options = D2D1_DRAW_TEXT_OPTIONS_NONE;
+    // D2D1_DRAW_TEXT_OPTIONS_ENABLE_COLOR_FONT enables a bunch of internal machinery
+    // which doesn't have to run if we know we can't use it anyways in the shader.
+    WI_SetFlagIf(options, D2D1_DRAW_TEXT_OPTIONS_ENABLE_COLOR_FONT, WI_IsFlagSet(value->flags, CellFlags::ColoredGlyph));
+
+    _r.d2dRenderTarget->BeginDraw();
+    // We could call
+    //   _r.d2dRenderTarget->PushAxisAlignedClip(&rect, D2D1_ANTIALIAS_MODE_ALIASED);
+    // now to reduce the surface that needs to be cleared, but this decreases
+    // performance by 10% (tested using debugGlyphGenerationPerformance).
+    _r.d2dRenderTarget->Clear();
+    _r.d2dRenderTarget->DrawTextLayout({}, textLayout.get(), _r.brush.get(), options);
+    THROW_IF_FAILED(_r.d2dRenderTarget->EndDraw());
+
+    for (uint32_t i = 0; i < cells; ++i)
+    {
+        // Specifying NO_OVERWRITE means that the system can assume that existing references to the surface that
+        // may be in flight on the GPU will not be affected by the update, so the copy can proceed immediately
+        // (avoiding either a batch flush or the system maintaining multiple copies of the resource behind the scenes).
+        //
+        // Since our shader only draws whatever is in the atlas, and since we don't replace glyph tiles that are in use,
+        // we can safely (?) tell the GPU that we don't overwrite parts of our atlas that are in use.

```

Comment:
>safely (?) [](http://example.com/codeflow?start=18&length=10)

the ? made me laugh

Reply:
Don't you know how people need to print "don't eat the plastic" on food because idiots would choke on it?
Well the D3D folks didn't do that. I'm choking.